storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -55,6 +113,2298 @@ public class UnifiedLog {
+        final AtomicLong diff = new AtomicLong(size() - retentionSize);
+
+        DeletionCondition shouldDelete = (segment, nextSegmentOpt) -> {
+            int segmentSize = segment.size();
+            boolean delete = diff.get() - segmentSize >= 0;
+            logger.debug("{} retentionSize breached: {}, log size before delete segment={}, after delete segment={}",
+                    segment, delete, diff.get(), diff.get() - segmentSize);
+            if (delete) {
+                diff.addAndGet(-segmentSize);
+            }
+            return delete;
+        };
+        return deleteOldSegments(shouldDelete, toDelete -> {
+            long size = size();
+            for (LogSegment segment : toDelete) {
+                size -= segment.size();
+                if (remoteLogEnabledAndRemoteCopyEnabled()) {
+                    logger.info("Deleting segment {} due to local log retention size {} breach. Local log size after deletion will be {}.",
+                            segment, UnifiedLog.localRetentionSize(config(), true), size);
+                } else {
+                    logger.info("Deleting segment {} due to log retention size {} breach. Log size after deletion will be {}.",
+                            segment, config().retentionSize, size);
+                }
+            }
+        });
+    }
+
+    private int deleteLogStartOffsetBreachedSegments() throws IOException {
+        DeletionCondition shouldDelete = (segment, nextSegmentOpt) -> {
+            boolean isRemoteLogEnabled = remoteLogEnabled();
+            long localLSO = localLogStartOffset();
+            long logStartOffsetValue = isRemoteLogEnabled ? localLSO : logStartOffset();
+            boolean delete = nextSegmentOpt
+                    .map(nextSegment -> nextSegment.baseOffset() <= logStartOffsetValue)
+                    .orElse(false);
+            logger.debug("{} logStartOffset breached: {}, nextSegmentOpt={}, {}",
+                    segment, delete, nextSegmentOpt, isRemoteLogEnabled ? "localLogStartOffset=" + localLSO : "logStartOffset=" + logStartOffset);
+            return delete;
+        };
+        return deleteOldSegments(shouldDelete, toDelete -> {
+            if (remoteLogEnabledAndRemoteCopyEnabled()) {
+                logger.info("Deleting segments due to local log start offset {} breach: {}",
+                        localLogStartOffset(), toDelete.stream().map(LogSegment::toString).collect(Collectors.joining(",")));
+            } else {
+                logger.info("Deleting segments due to log start offset {} breach: {}",
+                        logStartOffset, toDelete.stream().map(LogSegment::toString).collect(Collectors.joining(",")));
+            }
+        });
+    }
+
+    public boolean isFuture() {
+        return localLog.isFuture();
+    }
+
+    /**
+     * The size of the log in bytes
+     */
+    public long size() {
+        return localLog.segments().sizeInBytes();
```

Comment:
It's probably better to use `logSegments()` here.

Reply:
Good point, updated